optuna/_optimize.py

```diff
@@ -0,0 +1,248 @@
+
+            time_start = datetime.datetime.now()
+
+            def _should_stop() -> bool:
+                if study._stop_flag:
+                    return True
+
+                if timeout is not None:
+                    # This is needed for mypy.
+                    t = timeout  # type: float
+                    return (datetime.datetime.now() - time_start).total_seconds() > t
+
+                return False
+
+            if n_trials is not None:
+                _iter = iter(range(n_trials))
+            else:
+                _iter = iter(_should_stop, True)
+
+            with Parallel(n_jobs=n_jobs, prefer="threads") as parallel:
+                if not isinstance(
+                    parallel._backend, joblib.parallel.ThreadingBackend
+                ) and isinstance(study._storage, storages.InMemoryStorage):
+                    warnings.warn(
+                        "The default storage cannot be shared by multiple processes. "
+                        "Please use an RDB (RDBStorage) when you use joblib for "
+                        "multi-processing. The usage of RDBStorage can be found in "
+                        "https://optuna.readthedocs.io/en/stable/tutorial/rdb.html.",
+                        UserWarning,
+                    )
+
+                parallel(
+                    delayed(_optimize_sequential)(
+                        study,
+                        func,
+                        1,
+                        timeout,
+                        catch,
+                        callbacks,
+                        gc_after_trial,
+                        reseed_sampler_rng=True,
+                        time_start=time_start,
+                        progress_bar=None,
+                    )
+                    for _ in _iter
+                )
+    finally:
+        study._optimize_lock.release()
+        progress_bar.close()
+
+
+def _optimize_sequential(
+    study: "optuna.Study",
+    func: "optuna.study.ObjectiveFuncType",
+    n_trials: Optional[int],
+    timeout: Optional[float],
+    catch: Tuple[Type[Exception], ...],
+    callbacks: Optional[List[Callable[["optuna.Study", FrozenTrial], None]]],
+    gc_after_trial: bool,
+    reseed_sampler_rng: bool,
+    time_start: Optional[datetime.datetime],
+    progress_bar: Optional[pbar_module._ProgressBar],
+) -> None:
+    if reseed_sampler_rng:
+        study.sampler.reseed_rng()
+
+    i_trial = 0
+
+    if time_start is None:
+        time_start = datetime.datetime.now()
+
+    while True:
+        if study._stop_flag:
+            break
+
+        if n_trials is not None:
+            if i_trial >= n_trials:
+                break
+            i_trial += 1
+
+        if timeout is not None:
+            elapsed_seconds = (datetime.datetime.now() - time_start).total_seconds()
+            if elapsed_seconds >= timeout:
+                break
+
+        try:
+            trial = _run_trial(study, func, catch)
```

Comment:
`_run_trial` no longer cares about `gc`. It's handled here at the caller. Should reduce code complexity.